Opencast-Moodle/moodle-block_opencast
# Adopt video visibility after importing it with duplication method

This PR fixes #191
## How it works
After a duplicate event process is completed and the duplicate workflow has been started, a new adhoc task will be queued to adopt the visibility of the source video into the duplicated video.
This adhoc task waits for the duplication workflow process to be finished (at around 60 times with each an interval of 1 min in between), so, it checks every minute to proceeed with the visibility change.
- Changing visibility follows only the VISIBLE AND HIDDEN state of the source video, because GOUP visibility is not available in the new course.
- This task also give 5 tries to the failed attempts.
- As this feature is implemented and tested against opencast 13, therefore it is suggested to switch to default duplicate event workflow, if you don't have any custom workflow for this matter.

## New methods and classes:
- `block_opencast\task\process_duplicated_event_visibility_change` adhoc task under `classes/task`
- `set_duplicated_event_visibility()` method in `block_opencast\local\apibridge`
## To test this PR:
You need to have:
- Latest tool_opencast v4.0-r1 or higher
- A mandetory version upgrade for this plugin must take place in order to make the new adhoc task available to the moodle server.
- Duplicate event workflow is required to be set in the setting admin.
### Scenario to test this:
1. Review the admin setting and make sure every setting related to import videos is correctly set.
2. Set import mode as Duplicate Event
3. Make sure Workflow for duplicating events is set to duplicate-event **workflow**
4. Allow import is enabled.
5. Save changes
6. In a course to import videos from make sure you have set different visibilities for the videos (e.g. one visible to all and other one hidden and one with group visibility)
7. Try to import video (fast way is via manual import in overview page) (long way is to do it via backup and import a course)
8. Make sure cron job is up and running, specially for adhoc tasks.
9. After that the video are imported to the destination course/series and duplicate-event workflow is done, you need to also wait for the visibilities to get into place automatically.
10. A republish-metadata workflow will be run after all to set the correct visibility for the new imported events.
11. To check the visibilites in the new course, you have to make sure that:
 - if a video was visible to all in the origin course: It has to be visibile in the new course as well.
 - if the video was hidden or in a group: The video must be hidden in the new course
